README.md
# pscircle 

**pscircle** visualizes Linux processes in a form of radial tree.

*by Ruslan Kuchumov, 2018*

# Installation

## Compiling from the source code

Make sure you have [meson](http://mesonbuild.com/Getting-meson.html) and [ninja](https://ninja-build.org/) build systems installed.

You will also need the following dependencies:

*  libcairo
*  libpng
*  libx11 (optional; if disabled output only to PNG file will be supported)

In Debian-based distributions you can install them from the repository:

```bash
# Ninja and Meson build systems
sudo apt-get install -y ninja-build python3-pip python3-setuptools pkg-config
pip3 install meson

# Dependencies
sudo apt-get install -y libpng-dev libcairo2-dev libx11-dev
```

> In other distributions the process is similar, but the names of the packages may differ

To compile and install *pscircle* call:

```bash
mkdir build && cd build
meson ..
ninja
sudo ninja install
```

In case you want to compile without X11 support, call `meson configure -Denable-x11=false` before compiling.

## Installing from repositories

### ArchLinux AUR

A package is available on the AUR as
[pscircle-git](https://aur.archlinux.org/packages/pscircle-git/).

### Gentoo ebuild

Gentoo ebuild is available in [examples/pscircle-9999.ebuild](examples/pscircle-9999.ebuild) 
(by [_ahrs](https://www.reddit.com/r/unixporn/comments/8v9r19/oc_ive_just_created_a_program_for_drawing_process/e1qmuie)).

## Usage

After installation and configuration you may want to create systemd service to regularly update desktop wallpaper:

```bash
mkdir -p ~/.config/systemd/user/
cp examples/pscircle.service ~/.config/systemd/user/
systemctl --user enable pscircle
systemctl --user start pscircle
```

To view complete list of command line arguments run:

```bash
pscircle --help
```

Parameters related to visualization are described in the following diagram:

[![parameters](docs/parameters-small.png)](docs/parameters.png)

[full size](docs/parameters.png)

By default image will be printed directly to X11 root window of the display specified in `--output-display=...`. But you can print image to PNG file by specifying `--output=pscircle.png` argument. (Check [#6 example](examples/06-output-to-image.sh))

When you finish configuring command line arguments you can modify corresponding constants in [config.h.meson](config.h.meson) and recompile the application so that your configuration would be used by default.

## Using *pscircle* with desktop environments

GNOME, KDE, XFCE and, probably, some other DE do not display X11 root window as desktop wallpaper. But you can still save the image to the file (by using `--output=path.png`) and either:

* Set the path to the image in wallpaper system config. DE is likely to update the wallpaper automatically, when the image has changed.
* Create your own script for updating the wallpaper after *pscircle* is called. (I'd appreciate if you share it.)

> Hint: you can place the image in `/tmp` directory which is often mapped to the main memory. It omits writing to the disk and improves performance.

For GNOME DE [this script](examples/gnome.sh) may help you.

# Examples

For more examples check [examples](examples/) directory.

## Default

[![default](docs/default-small.png)](docs/default.png)

[full size](docs/default.png)

## Tree at the left

[![tree at the left](docs/01-left-small.png)](docs/parameters.png)

[source](examples/01-left.sh) | [full size](docs/parameters.png)

## Different fonts and colors

[![Different fonts and colors](docs/02-colors-small.png)](docs/02-colors.png)

[source](examples/02-colors.sh) | [full size](docs/02-colors.png)

## Background image

[![Background image](docs/03-background-small.png)](docs/03-background.png)

[source](examples/03-background.sh) | [full size](docs/03-background.png)

## Tree at the bottom

[![tree at the left](docs/04-bottom-small.png)](docs/04-bottom.png)

[source](examples/04-bottom.sh) | [full size](docs/04-bottom.png)

# Rationale 

## Graph doesn't fit to my screen

By default *pscircle* creates an image for 3200x1800 resolution. If you have a different screen resolution, you should manually specify it by changing `--output-width` and `--output-height` options. You will probably need to change other options related to sizes (e.g. font size, circles radii, lists positions) as their values are specified in absolute values and do not scale automatically. [Example #5](examples/05-1440x900-res.sh) may help you with this.

Another reason the graph doesn't fit to the screen is that you have too much processes. In this case you can reduce circle radii (by changing `--tree-radius-increment` option), limit the number of child processes (`--max-children` option) or change the PID of the root procces (`--root-pid` option). You can also hide lists of the processes with `--cpulist-show=false` or `--memlist-show=false` to get more free space at the screen.

## How does *pscircle* obtain its data?

*pscircle* reads `/proc` Linux pseudo file system for collecting information about processes, uptime, CPU and memory utilization. It calculates processes pcpu and CPU utilization values over the time interval specified in `--interval` argument. This implies that *pscircle* execution is suspended for this time interval to collect the data. If you want these values to be calculated from the process and system start time (similar to `top` or `htop` utils), you can specify `--interval=0` argument.

In case *pscircle* doesn't work correctly with your kernel version (please, let me know), or you want to monitor remote host you can provide this information yourself. (Check [example #7](examples/07-no-proc-fs.sh))

## Performance

When you compile *pscircle* with `#define PSC_PRINT_TIME 1` cpu time and wall time for different stages will be printed:

Execution times of printing the image to X11 root window:

```
                  cpu /   wall 
         init: 0.0292 / 1.0296 seconds
      arrange: 0.0003 / 0.0003 seconds
    draw tree: 0.0844 / 0.1443 seconds
   draw lists: 0.0009 / 0.0031 seconds
        write: 0.0000 / 0.0000 seconds
        total: 0.1149 / 1.1772 seconds
```

Execution times of printing the image to PNG file:

```
                  cpu /   wall 
         init: 0.0199 / 1.0200 seconds
      arrange: 0.0001 / 0.0001 seconds
    draw tree: 0.1062 / 0.1062 seconds
   draw lists: 0.0009 / 0.0009 seconds
        write: 0.3211 / 0.3211 seconds
        total: 0.4481 / 1.4483 seconds
```

As you can see, drawing the tree to the file (on SSD disk) take almost 4 times longer that printing to X11 screen (0.11 vs 0.45 seconds of cputime). 

## Multiple display environment

As *pscircle* is not tested yet in multi-display environment to make it work correctly, I suggest trying the following options:

* specify `--output-display=...` argument to print the image to the correct display, or
* change `DISPLAY` environment variable before running the program, or
* output image to the file using `--output=...` argument and then set it as the wallpaper using external command (e.g. `feh`).

# Contributing

You can suggest your desired features [here](https://gitlab.com/mildlyparallel/pscircle/issues/7) and vote for the features suggested by other users.

# Asknowlegemnts

Thanks Bill MIll for writing an article on [tree visualization algorithm](https://llimllib.github.io/pymag-trees/) which was used in this program.


CHANGELOG.md
# 1.2 - 2018.07.20

## Added

* Multiple circles radii support (e.g `--tree-radius-increment=30,40,50`)
* Option for hiding tree levels (`--hide-top-levels=1`)
* GNOME example
* `--version` option

# 1.1 - 2018.07.06

## Added

* `--collapse-threads` option
* Gentoo ebuild
* Pywal script

## Updated

* Readme: descriptions and guides, fixed typos

## Removed

* `-mavx` compilation option,


src/cfg.c
#include <assert.h>
#include "argparser.h"

#include <stdlib.h>
#include <stdio.h>
#include "cfg.h"

#ifndef M_PI
#define M_PI R(3.14159265358979323846)
#endif

#define rgb(r,g,b) {(real_t)(r)/255, (real_t)(g)/255, (real_t)(b)/255, 1}

#define rgba(r,g,b,a) {(real_t)(r)/255, (real_t)(g)/255, (real_t)(b)/255, (real_t)(a)/255}

#define list(...) { .d = { __VA_ARGS__ } }

#include "config.h"

cfg_t config = {
	.read_stdin = PSC_STDIN,
	.interval   = PSC_INTERVAL,

	.output           = PSC_OUTPUT,
	.output_width     = PSC_OUTPUT_WIDTH,
	.output_height    = PSC_OUTPUT_HEIGHT,
	.output_display   = PSC_OUTPUT_DISPLAY,
	.memory_unit      = PSC_MEMORY_UNIT,
	.root_pid         = PSC_ROOT_PID,
	.max_children     = PSC_MAX_CHILDREN,
	.collapse_threads = PSC_COLLAPSE_THREADS,
	.background       = PSC_BACKGROUND_COLOR,
	.background_image = PSC_BACKGROUND_IMAGE,

	.max_mem = PSC_MEM_MAX,
	.min_mem = PSC_MEM_MIN,
	.max_cpu = PSC_CPU_MAX,
	.min_cpu = PSC_CPU_MIN,

	.tree = {
		.font_face  = PSC_TREE_FONT_FACE,
		.font_size  = PSC_TREE_FONT_SIZE,
		.font_color = PSC_TREE_FONT_COLOR,
		.radius_inc = PSC_TREE_RADIUS_INCREMENT,
		.sector     = PSC_TREE_SECTOR,
		.rotate     = PSC_TREE_ROTATE,
		.rotation   = PSC_TREE_ROTATION,
		.center     = PSC_TREE_CENTER,
		.anchor_proc_name  = PSC_ANCHOR_PROC_NAME,
		.anchor_proc_angle = PSC_ANCHOR_PROC_ANGLE,
	},

	.dot = {
		.radius = PSC_DOT_RADIUS,
		.border = PSC_DOT_BORDER,
		.fg_min = PSC_DOT_BORDER_COLOR_MIN,
		.fg_max = PSC_DOT_BORDER_COLOR_MAX,
		.bg_min = PSC_DOT_BACKGROUND_COLOR_MIN,
		.bg_max = PSC_DOT_BACKGROUND_COLOR_MAX,
	},

	.link = {
		.width     = PSC_LINK_WIDTH,
		.convexity = PSC_LINK_CONVIXITY,
		.color_min = PSC_LINK_COLOR_MIN,
		.color_max = PSC_LINK_COLOR_MAX
	},

	.toplists = {
		.row_height     = PSC_TOPLISTS_ROW_HEIGHT,
		.font_size      = PSC_TOPLISTS_FONT_SIZE,
		.font_color     = PSC_TOPLISTS_FONT_COLOR,
		.font_face      = PSC_TOPLISTS_FONT_FACE,
		.pid_font_color = PSC_TOPLISTS_PID_FONT_COLOR,
		.column_padding = PSC_TOPLISTS_COLUMN_PADDING,

		.bar = {
			.width      = PSC_TOPLISTS_BAR_WIDTH,
			.height     = PSC_TOPLISTS_BAR_HEIGHT,
			.background = PSC_TOPLISTS_BAR_BG,
			.color      = PSC_TOPLISTS_BAR_COLOR,
		},

		.cpulist = {
			.show         = PSC_CPULIST_SHOW,
			.center       = PSC_CPULIST_CENTER,
			.show_header  = PSC_CPULIST_SHOW_HEADER,
			.name         = PSC_CPULIST_NAME,
			.label        = PSC_CPULIST_LABEL,
			.value_format = PSC_CPULIST_VALUE_FMT,
			.value        = PSC_CPULIST_BAR_VALUE,
		},

		.memlist = {
			.show         = PSC_MEMLIST_SHOW,
			.center       = PSC_MEMLIST_CENTER,
			.show_header  = PSC_MEMLIST_SHOW_HEADER,
			.name         = PSC_MEMLIST_NAME,
			.label        = PSC_MEMLIST_LABEL,
			.value_format = PSC_MEMLIST_VALUE_FMT,
			.value        = PSC_MEMLIST_BAR_VALUE,
		},

	}
};

void
parse_cmdline(int argc, char const * argv[])
{
	assert(argv);
	assert(argc >= 1);

	if (argc == 1)
		return;

	argparser_t argp = {0};
	argparser_init(&argp);

	ARGQ(&argp, "--stdin", config.read_stdin, parser_bool, PSC_STDIN,
		"If set to true, process data will be read from stdin in form of"
		"`ps -e -o pid,ppid,pcpu,rss,comm --no-headers` output. Otherwise, "
		"/proc file system will be read to obtain the list of processes and information "
		"on memory usage, load average and CPU utilization");
	ARGQ(&argp, "--interval", config.interval, parser_real, PSC_INTERVAL,
		"If set to 0 (default), CPU utilization and processes PCPU values will be calculate "
		"from system start time and proceess start time. Otherwise, these values will be calculated "
		"over specified interval (in seconds, with fractions). This also implies that program exection "
		"will be suspended to the specified interval.");
#ifdef HAVE_X11
	ARG(&argp, "--output", config.output, parser_string, PSC_OUTPUT,
		"Path to the output image. If it's not set, X11 root window is used");
	ARG(&argp, "--output-display", config.output_display, parser_string, PSC_OUTPUT_DISPLAY,
		"Name of X11 display to draw the image to");
#else
	ARG(&argp, "--output", config.output, parser_string, PSC_OUTPUT,
		"Path to the output image");
#endif
	ARGQ(&argp, "--output-width", config.output_width, parser_ulong, PSC_OUTPUT_WIDTH,
		"Width(px) of output image or X11 root window");
	ARGQ(&argp, "--output-height", config.output_height, parser_ulong, PSC_OUTPUT_HEIGHT,
		"Height(px) of output image or X11 root window");

	ARGQ(&argp, "--root-pid", config.root_pid, parser_long, PSC_ROOT_PID,
		"PID of the root process");
	ARGQ(&argp, "--max-children", config.max_children, parser_long, PSC_MAX_CHILDREN,
		"Maximum number of child proceceses.");
	ARGQ(&argp, "--collapse-threads", config.collapse_threads, parser_bool, PSC_COLLAPSE_THREADS,
		"If set, processes children with the name and without their own children will be collpsed into "
		"a single proccess");

	ARGQ(&argp, "--memory-unit", config.memory_unit, parser_memory_unit, PSC_MEMORY_UNIT,
		"Unit of memeory (B, K, M, G, T) used in RSS memory column");
	ARGQ(&argp, "--memory-min-value", config.min_mem, parser_memory, PSC_MEM_MIN,
			"Processes with RSS below specified value will have --dot-border-color-min "
			"and --link-color-min color");
	ARGQ(&argp, "--memory-max-value", config.max_mem, parser_memory, PSC_MEM_MAX,
			"Processes with RSS above specified value will have --dot-border-color-max "
			"and --link-color-max color");
	ARGQ(&argp, "--cpu-min-value", config.min_cpu, parser_real, PSC_CPU_MIN,
			"Processes with PCPU below specified value will have --dot-color-min color");
	ARGQ(&argp, "--cpu-max-value", config.max_cpu, parser_real, PSC_CPU_MAX,
			"Processes with PCPU above specified value will have --dot-color-max color");

	ARG(&argp, "--background-color", config.background, parser_color, color_to_hex((color_t) PSC_BACKGROUND_COLOR),
			"Image backgound color");
	ARG(&argp, "--background-image", config.background_image, parser_string, PSC_BACKGROUND_IMAGE,
			"Path to background image. Image will be drawn at the top left corner without scaling");

	ARG(&argp, "--tree-center", config.tree.center, parser_point, point_to_str((point_t) PSC_TREE_CENTER),
			"X:Y Position of a tree center from the center of image");
	ARG(&argp, "--tree-radius-increment", config.tree.radius_inc, parser_list_real, reals_to_string(PSC_TREE_RADIUS_INCREMENT),
			"The diffrence between radii of concentric circiles of the tree. Multiple values "
			"seprated with comma can be specified, e.g --tree-radius-increment=30,40,50 "
			"would mean that the first radius would be 30, the second 70, and the reset would "
			"be incremented by 50 (120,170,220...)");
	ARGQ(&argp, "--hide-top-levels", config.tree.hide_levels, parser_ulong, PSC_HIDE_TOP_LEVELS,
			"Specified number of the tree levels from the top will not be displayed.");
	ARGQ(&argp, "--tree-sector-angle", config.tree.sector, parser_real, PSC_TREE_SECTOR,
			"Tree vertices will be displayed inside the sector with specified angle");
	ARGQ(&argp, "--tree-rotate", config.tree.rotate, parser_bool, PSC_TREE_ROTATE,
			"Tree will be rotated to the angle specified in --tree-rotation-angle"
			"If this option and --tree-anchor-proc-name are not set, "
			"then the tree will be rotated to so that its widest node is at 0 rad");
	ARGQ(&argp, "--tree-rotation-angle", config.tree.rotation, parser_real, PSC_TREE_ROTATION,
			"Rotation angle of a tree (radians). If --tree-rotate is not set, the tree will "
			"be rotated to --tree-anchor-proc-angle (if --tree-anchor-proc-name is set) or to "
			"the angle of the widest child");
	ARG(&argp, "--tree-anchor-proc-name", config.tree.anchor_proc_name, parser_string, PSC_ANCHOR_PROC_NAME,
			"Tree will be rotated so that specified process is positioned at --tree-anchor-proc-angle angle"
			", unless --tree-rotate is set. If this option and --tree-rotate are not set or the process is not found, "
			"then the tree will be rotated to so that its widest node is at 0 rad");
	ARGQ(&argp, "--tree-anchor-proc-angle", config.tree.anchor_proc_angle, parser_real, PSC_ANCHOR_PROC_ANGLE,
			"Tree will be rotated so that proccess specified by --tree-anchor-proc-name is positioned "
			"the specified angle, unless --tree-rotate is set.");
	ARGQ(&argp, "--tree-font-size", config.tree.font_size, parser_real, PSC_TREE_FONT_SIZE,
			"Font size of the tree process names");
	ARG(&argp, "--tree-font-face", config.tree.font_face, parser_string, PSC_TREE_FONT_FACE,
			"Font face of the tree process names");
	ARG(&argp, "--tree-font-color", config.tree.font_color, parser_color, color_to_hex((color_t) PSC_TREE_FONT_COLOR),
			"The color of the tree process names");

	ARGQ(&argp, "--dot-radius", config.dot.radius, parser_real, PSC_DOT_RADIUS,
			"Radius of the dots (px)");
	ARGQ(&argp, "--dot-border-width", config.dot.border, parser_real, PSC_DOT_BORDER,
			"Width of dots borders (px)");
	ARG(&argp, "--dot-color-min", config.dot.bg_min, parser_color, color_to_hex((color_t) PSC_DOT_BACKGROUND_COLOR_MIN),
			"Backgound color of the dots. This value corresponds to --cpu-max-value");
	ARG(&argp, "--dot-color-max", config.dot.bg_max, parser_color, color_to_hex((color_t) PSC_DOT_BACKGROUND_COLOR_MAX),
			"Backgound color of the dots. This value corresponds to --cpu-max-value");
	ARG(&argp, "--dot-border-color-min", config.dot.fg_min, parser_color, color_to_hex((color_t) PSC_DOT_BORDER_COLOR_MIN),
			"Border color of the dots. This value corresponds to --memory-min-value");
	ARG(&argp, "--dot-border-color-max", config.dot.fg_max, parser_color, color_to_hex((color_t) PSC_DOT_BORDER_COLOR_MAX),
			"Border color of the dots. This value corresponds to --memory-max-value");

	ARGQ(&argp, "--link-width", config.link.width, parser_real, PSC_LINK_WIDTH,
			"Width of the curves between dots");
	ARGQ(&argp, "--link-convexity", config.link.convexity, parser_real, PSC_LINK_CONVIXITY,
			"Convexity of links curves. Curve control points are offset from the "
			"centers of the dots to the distance of --tree-radius-increment times "
			"this value");
	ARG(&argp, "--link-color-min", config.link.color_min, parser_color, color_to_hex((color_t) PSC_LINK_COLOR_MIN),
			"Color of the curves betwwen dots. this value corresponds to --memory-min-value");
	ARG(&argp, "--link-color-max", config.link.color_max, parser_color, color_to_hex((color_t) PSC_LINK_COLOR_MAX),
			"Color of the curves betwwen dots. this value corresponds to --memory-max-value");

	ARGQ(&argp, "--toplists-row-height", config.toplists.row_height, parser_real, PSC_TOPLISTS_ROW_HEIGHT,
			"Hight of each row in toplist (px)");
	ARGQ(&argp, "--toplists-font-size", config.toplists.font_size, parser_real, PSC_TOPLISTS_FONT_SIZE,
			"Font size of the text in toplists");
	ARG(&argp, "--toplists-font-color", config.toplists.font_color, parser_color, color_to_hex((color_t) PSC_TOPLISTS_FONT_COLOR),
			"Font color of the text in toplists");
	ARG(&argp, "--toplists-pid-font-color", config.toplists.pid_font_color, parser_color, color_to_hex((color_t) PSC_TOPLISTS_PID_FONT_COLOR),
			"Font of processes PIDs in toplists");
	ARG(&argp, "--toplists-font-face", config.toplists.font_face, parser_string, PSC_TOPLISTS_FONT_FACE,
			"Font of the text in toplists");
	ARGQ(&argp, "--toplists-column-padding", config.toplists.column_padding, parser_real, PSC_TOPLISTS_COLUMN_PADDING,
			"Font of the text in toplists");

	ARGQ(&argp, "--toplists-bar-width", config.toplists.bar.width, parser_real, PSC_TOPLISTS_BAR_WIDTH,
			"Width of the percentage bar and legend markers in each row");
	ARGQ(&argp, "--toplists-bar-height", config.toplists.bar.height, parser_real, PSC_TOPLISTS_BAR_HEIGHT,
			"Height of the percentage bar");
	ARG(&argp, "--toplists-bar-background", config.toplists.bar.background, parser_color, color_to_hex((color_t) PSC_TOPLISTS_BAR_BG),
			"Backgound color of the percentage bar");
	ARG(&argp, "--toplists-bar-color", config.toplists.bar.color, parser_color, color_to_hex((color_t) PSC_TOPLISTS_BAR_COLOR),
			"Foreground color of the percentage bar");

	ARGQ(&argp, "--cpulist-show", config.toplists.cpulist.show, parser_bool, PSC_CPULIST_SHOW,
			"Shows the list of porcesses with max CPU utilization");
	ARG(&argp, "--cpulist-center", config.toplists.cpulist.center, parser_point, point_to_str((point_t) PSC_CPULIST_CENTER),
			"Position of the center of CPU list");
	ARGQ(&argp, "--cpulist-show-header", config.toplists.cpulist.show_header, parser_bool, PSC_CPULIST_SHOW_HEADER,
			"Show the header of the CPU toplist");
	ARG(&argp, "--cpulist-name", config.toplists.cpulist.name, parser_string, PSC_CPULIST_NAME,
			"The string to the left of the progress bar of CPU toplist");
	ARG(&argp, "--cpulist-label", config.toplists.cpulist.label, parser_string, PSC_CPULIST_LABEL,
			"The string to the right of the progress bar of CPU toplist");
	ARGQ(&argp, "--cpulist-bar-value", config.toplists.cpulist.value, parser_real, PSC_CPULIST_BAR_VALUE,
			"The value (0<=x<=1) of percentage bar of CPU toplist");

	ARGQ(&argp, "--memlist-show", config.toplists.memlist.show, parser_bool, PSC_MEMLIST_SHOW,
			"Shows the list of porcesses with max memory utilization");
	ARG(&argp, "--memlist-center", config.toplists.memlist.center, parser_point, point_to_str((point_t) PSC_MEMLIST_CENTER),
			"Position of the center of MEM list");
	ARGQ(&argp, "--memlist-show-header", config.toplists.memlist.show_header, parser_bool, PSC_MEMLIST_SHOW_HEADER,
			"Show the header of the MEM toplist");
	ARG(&argp, "--memlist-name", config.toplists.memlist.name, parser_string, PSC_MEMLIST_NAME,
			"The string to the left of the progress bar of MEM toplist");
	ARG(&argp, "--memlist-label", config.toplists.memlist.label, parser_string, PSC_MEMLIST_LABEL,
			"The string to the right of the progress bar of MEM toplist");
	ARGQ(&argp, "--memlist-bar-value", config.toplists.memlist.value, parser_real, PSC_MEMLIST_BAR_VALUE,
			"The value (0<=x<=1) of percentage bar of MEM toplist");

	argparser_parse(&argp, argc, argv);
}

